### DISCUSSION TITLE
Problems with sequent smart fan, new hardware revision

### DISCUSSION THREAD

**Discussion by lockeandev** [tagyoureit/nodejs-poolController#593] [REM]
New discussion to troubleshoot hardware version 4.0 of the sequent smart fan and integration with remchem and njsPC. 

The new revision does not report fan temp correctly when units are set to Celsius.

**Comment by rstrouse** [comment]
I do not have this device so can you tell me what is displayed in REM for the Hardware version?

**Reply by tagyoureit** [reply]
This convo started here: https://github.com/tagyoureit/nodejs-poolController/discussions/502#discussioncomment-3542640

I had @lockeandev start a new thread so we could track it more easily, but then dropped the ⚽ ball completely.  :)

**Reply by lockeandev** [reply]
The hardware and firmware versions in rem read as 129.28.
![Screenshot_20220909_102156_Chrome](https://user-images.githubusercontent.com/11232490/189396706-fa4b74bd-f7a5-432d-a8f3-00cc2c919da8.jpg)

**Reply by rstrouse** [reply]
Well it looks like they came up short on cpu supply so created the version 4 hardware which in itself does not report the hardware version.  The way they know the hardware version is based upon the address.  Hardware versions < 4 will be at i2C address 3 (0x03) or 4 (0x04) and hardware version 4 is at address 44 (0x2c) or 45 (0x2d).   It does appear that all the functions are the same (sans version reporting) but the chip registers are different for some of the calls.

The other thing this version does is occupy a GPIO pin (BCM12) or pin 32 and you cannot set the safety temp via the card.  There is also no Fan Temp, blinky light, or timed fan operation.  I think they have removed or changed the internal micro processor.  I don't know if any of those features will be missed or not.

**Reply by tagyoureit** [reply]
From his other post I linked above, the command line returns the right version.  `I will have to try to take a picture later but running the command line tool fan -list: board returns hardware version 4.0`.  We should be able to figure it out from something besides the i2c address.  Or are you say that _is_ what they are doing?

**Reply by rstrouse** [reply]
The text hardware version 4.0 is actually hardcoded.  The InitBoard function tries to find the board at 0x3 or 0x4 and if it fails it hardcodes 4.0 as the version.  So yes we should simply look at the address and if it is > 0x04 then it is version 4.  Then in the code there are a bunch of places where it encapsulates the `if ver < 4` then either blow off the call or call something else.   Some of the functions as listed above are not supported on 4+.

**Reply by tagyoureit** [reply]
Mihai is cutting too many corners.  Boo!

**Reply by Telemachus1977** [reply]
Experienced similar problems with a fan board that my buddy bought.

Functionality is quite limited compared to those mentioned in the wiki.  Also the fan squeaks out of the box.

**Reply by rock-crusher** [reply]
> Mihai is cutting too many corners. Boo!



> Experienced similar problems with a fan board that my buddy bought.
> 
> Functionality is quite limited compared to those mentioned in the wiki. Also the fan squeaks out of the box.

Yep, if they keep that up… it’s the beginning of the end.
People will seek other alternative products to suit their needs.
Quality control is another problem; unfortunately, it seems many companies will simply use the general public for that.

**Comment by mhmatttu** [comment]
curious if this was ever resolved.  i just recieved this fan...v6.0 printed on the board.  i can control it using the Sequent program but each time i add it to REM, it crashes REM and have to fully remove and reinstall REM.  I'm cautions to start stacking other cards and setting up in fear that any attempt to re-add the smart fan will result in a start over.

**Reply by tagyoureit** [reply]
This PR https://github.com/rstrouse/relayEquipmentManager/pull/65 was accepted for the SmartFan v6.

**Reply by FlaMike** [reply]
I just received and installed in the Pi stack a V6 Smart Fan.  In order to obtain the PR mentioned above, I updated REM.  However, REM will not restart after the update has been completed. PM2 says it's running but its web page will not load & NJS-PC and dashPanel are not happy.

If I try to do a minimal configuration of the fan in REM without the PR installed (identify the device, give it a name, set the temperature to F), REM crashes & I have to reboot.  I does not crash if I use C instead of F.  

I am running REM version 8.0.0

Do you have any suggestions on how to move forward?

**Reply by rstrouse** [reply]
Is there anything in the PM2 log for REM.  My guess is that there is probably some default setting for the V6 fan that I do not have that isn't being set.

**Reply by FlaMike** [reply]
Here are the last 150 lines from the log:

'
Error: Cannot find module 'ip'
Require stack:
- /home/michael/relayEquipmentManager/node_modules/node-ssdp/lib/index.js
- /home/michael/relayEquipmentManager/node_modules/node-ssdp/lib/server.js
- /home/michael/relayEquipmentManager/node_modules/node-ssdp/index.js
- /home/michael/relayEquipmentManager/dist/web/Server.js
- /home/michael/relayEquipmentManager/dist/i2c-bus/I2cBus.js
- /home/michael/relayEquipmentManager/dist/connections/Bindings.js
- /home/michael/relayEquipmentManager/dist/boards/Controller.js
- /home/michael/relayEquipmentManager/dist/app.js
    at Module._resolveFilename (node:internal/modules/cjs/loader:1134:15)
    at Module._load (node:internal/modules/cjs/loader:975:27)
    at Module.require (node:internal/modules/cjs/loader:1225:19)
    at require (node:internal/modules/helpers:177:18)
    at Object.<anonymous> (/home/michael/relayEquipmentManager/node_modules/node-ssdp/lib/index.js:6:10)
    at Module._compile (node:internal/modules/cjs/loader:1356:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1414:10)
    at Module.load (node:internal/modules/cjs/loader:1197:32)
    at Module._load (node:internal/modules/cjs/loader:1013:12)
    at Module.require (node:internal/modules/cjs/loader:1225:19) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/home/michael/relayEquipmentManager/node_modules/node-ssdp/lib/index.js',
    '/home/michael/relayEquipmentManager/node_modules/node-ssdp/lib/server.js',
    '/home/michael/relayEquipmentManager/node_modules/node-ssdp/index.js',
    '/home/michael/relayEquipmentManager/dist/web/Server.js',
    '/home/michael/relayEquipmentManager/dist/i2c-bus/I2cBus.js',
    '/home/michael/relayEquipmentManager/dist/connections/Bindings.js',
    '/home/michael/relayEquipmentManager/dist/boards/Controller.js',
    '/home/michael/relayEquipmentManager/dist/app.js'
  ]
}

Node.js v18.19.0
node:internal/modules/cjs/loader:1137
  throw err;
  ^

Error: Cannot find module 'ip'
Require stack:
- /home/michael/relayEquipmentManager/node_modules/node-ssdp/lib/index.js
- /home/michael/relayEquipmentManager/node_modules/node-ssdp/lib/server.js
- /home/michael/relayEquipmentManager/node_modules/node-ssdp/index.js
- /home/michael/relayEquipmentManager/dist/web/Server.js
- /home/michael/relayEquipmentManager/dist/i2c-bus/I2cBus.js
- /home/michael/relayEquipmentManager/dist/connections/Bindings.js
- /home/michael/relayEquipmentManager/dist/boards/Controller.js
- /home/michael/relayEquipmentManager/dist/app.js
    at Module._resolveFilename (node:internal/modules/cjs/loader:1134:15)
    at Module._load (node:internal/modules/cjs/loader:975:27)
    at Module.require (node:internal/modules/cjs/loader:1225:19)
    at require (node:internal/modules/helpers:177:18)
    at Object.<anonymous> (/home/michael/relayEquipmentManager/node_modules/node-ssdp/lib/index.js:6:10)
    at Module._compile (node:internal/modules/cjs/loader:1356:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1414:10)
    at Module.load (node:internal/modules/cjs/loader:1197:32)
    at Module._load (node:internal/modules/cjs/loader:1013:12)
    at Module.require (node:internal/modules/cjs/loader:1225:19) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/home/michael/relayEquipmentManager/node_modules/node-ssdp/lib/index.js',
    '/home/michael/relayEquipmentManager/node_modules/node-ssdp/lib/server.js',
    '/home/michael/relayEquipmentManager/node_modules/node-ssdp/index.js',
    '/home/michael/relayEquipmentManager/dist/web/Server.js',
    '/home/michael/relayEquipmentManager/dist/i2c-bus/I2cBus.js',
    '/home/michael/relayEquipmentManager/dist/connections/Bindings.js',
    '/home/michael/relayEquipmentManager/dist/boards/Controller.js',
    '/home/michael/relayEquipmentManager/dist/app.js'
  ]
}

Node.js v18.19.0
node:internal/modules/cjs/loader:1137
  throw err;
  ^

Error: Cannot find module 'ip'
Require stack:
- /home/michael/relayEquipmentManager/node_modules/node-ssdp/lib/index.js
- /home/michael/relayEquipmentManager/node_modules/node-ssdp/lib/server.js
- /home/michael/relayEquipmentManager/node_modules/node-ssdp/index.js
- /home/michael/relayEquipmentManager/dist/web/Server.js
- /home/michael/relayEquipmentManager/dist/i2c-bus/I2cBus.js
- /home/michael/relayEquipmentManager/dist/connections/Bindings.js
- /home/michael/relayEquipmentManager/dist/boards/Controller.js
- /home/michael/relayEquipmentManager/dist/app.js
    at Module._resolveFilename (node:internal/modules/cjs/loader:1134:15)
    at Module._load (node:internal/modules/cjs/loader:975:27)
    at Module.require (node:internal/modules/cjs/loader:1225:19)
    at require (node:internal/modules/helpers:177:18)
    at Object.<anonymous> (/home/michael/relayEquipmentManager/node_modules/node-ssdp/lib/index.js:6:10)
    at Module._compile (node:internal/modules/cjs/loader:1356:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1414:10)
    at Module.load (node:internal/modules/cjs/loader:1197:32)
    at Module._load (node:internal/modules/cjs/loader:1013:12)
    at Module.require (node:internal/modules/cjs/loader:1225:19) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/home/michael/relayEquipmentManager/node_modules/node-ssdp/lib/index.js',
    '/home/michael/relayEquipmentManager/node_modules/node-ssdp/lib/server.js',
    '/home/michael/relayEquipmentManager/node_modules/node-ssdp/index.js',
    '/home/michael/relayEquipmentManager/dist/web/Server.js',
    '/home/michael/relayEquipmentManager/dist/i2c-bus/I2cBus.js',
    '/home/michael/relayEquipmentManager/dist/connections/Bindings.js',
    '/home/michael/relayEquipmentManager/dist/boards/Controller.js',
    '/home/michael/relayEquipmentManager/dist/app.js'
  ]
}

Node.js v18.19.0
node:internal/modules/cjs/loader:1137
  throw err;
  ^

Error: Cannot find module 'ip'
Require stack:
- /home/michael/relayEquipmentManager/node_modules/node-ssdp/lib/index.js
- /home/michael/relayEquipmentManager/node_modules/node-ssdp/lib/server.js
- /home/michael/relayEquipmentManager/node_modules/node-ssdp/index.js
- /home/michael/relayEquipmentManager/dist/web/Server.js
- /home/michael/relayEquipmentManager/dist/i2c-bus/I2cBus.js
- /home/michael/relayEquipmentManager/dist/connections/Bindings.js
- /home/michael/relayEquipmentManager/dist/boards/Controller.js
- /home/michael/relayEquipmentManager/dist/app.js
    at Module._resolveFilename (node:internal/modules/cjs/loader:1134:15)
    at Module._load (node:internal/modules/cjs/loader:975:27)
    at Module.require (node:internal/modules/cjs/loader:1225:19)
    at require (node:internal/modules/helpers:177:18)
    at Object.<anonymous> (/home/michael/relayEquipmentManager/node_modules/node-ssdp/lib/index.js:6:10)
    at Module._compile (node:internal/modules/cjs/loader:1356:14)
    at Module._extensions..js (node:internal/modules/cjs/loader:1414:10)
    at Module.load (node:internal/modules/cjs/loader:1197:32)
    at Module._load (node:internal/modules/cjs/loader:1013:12)
    at Module.require (node:internal/modules/cjs/loader:1225:19) {
  code: 'MODULE_NOT_FOUND',
  requireStack: [
    '/home/michael/relayEquipmentManager/node_modules/node-ssdp/lib/index.js',
    '/home/michael/relayEquipmentManager/node_modules/node-ssdp/lib/server.js',
    '/home/michael/relayEquipmentManager/node_modules/node-ssdp/index.js',
    '/home/michael/relayEquipmentManager/dist/web/Server.js',
    '/home/michael/relayEquipmentManager/dist/i2c-bus/I2cBus.js',
    '/home/michael/relayEquipmentManager/dist/connections/Bindings.js',
    '/home/michael/relayEquipmentManager/dist/boards/Controller.js',
    '/home/michael/relayEquipmentManager/dist/app.js'
  ]
}

Node.js v18.19.0

'

**Reply by FlaMike** [reply]
And this is what pm2 shows.  I need to reload a backup because REM won't come up so the pool is down again :-(

![image](https://github.com/user-attachments/assets/23ec533e-eb81-4a54-a88b-341ac881773d)

**Reply by rstrouse** [reply]
Run the following commands

`pm2 stop REM`
`cd ~/relayEquipmentManager`
`git pull`
`npm i`
`npm start`


This will display logs.  It may be that your pm2 scripts are not compiling the typescript.  If there are no errors then hit CTRL+C and run `pm2 restart REM`

**Comment by FlaMike** [comment]
Ok.  Very similar to what I've tried, but will follow exactly & let you
know tomorrow.

Heat index is currently about 105, the pool is a chilly 90 & I need to get
in a few laps before happy hour

**Reply by rstrouse** [reply]
That's cold.  You are going to need a parka.

![image](https://github.com/user-attachments/assets/caeaaa96-668e-49b7-97c3-634e59ca7b72)

**Reply by FlaMike** [reply]
Fortunately, I had a drysuit so all was well :-)

**Reply by FlaMike** [reply]
Below are screenshots to show you exactly the commands entered and the system's response.  When all is completed, pm2 says REM has restarted, but it does not come back up--REM's web page is not reachable and oddly the poolController  dashboard is still active.  However in order to get REM going again, I have to reload everything from a backup SD card.

![image](https://github.com/user-attachments/assets/8b221d0d-495b-4f4f-ae9a-646b5601e403)

![image](https://github.com/user-attachments/assets/104fe9c6-7f1d-4f4a-860f-c27600faaa6f)

![image](https://github.com/user-attachments/assets/3746cd80-d143-40e0-8715-912beb50b0b6)

![image](https://github.com/user-attachments/assets/da7d0e2b-437e-44db-ac9a-f5dbcb84bb0d)


REM's web page after the "restart"

![image](https://github.com/user-attachments/assets/474bb434-70eb-43ab-b593-84031dedcb81)

**Comment by rstrouse** [comment]
`npm audit -fix`! Please do not run that.  Most of the time this has deleterious effects.  It is indiscriminate and assumes every device is a cloud server with an exposure plane to Venus.

This looks like the typescript compiler no longer works to build the node_modules directory. 
`cd ~/relayEquipmentManager`
`git reset --hard`
`git pull`
`rm -rf node_modules`
`npm i`
-- Do not run npm audit fix!!!
`npm start`

**Reply by FlaMike** [reply]
OK.  I've been getting the git pull error regarding package-lock.json & package.json files.  Do I just rename the files as .bkp & the proceed with the git pull?

***Never Mind!  I see the git reset --hard command in your instructions.  I'll just skip the audit fix.****

**Reply by FlaMike** [reply]
It looks like REM is running again!  Yay!  I now have a Smart Fan V6+ option on REM :-)

Do you have any suggestions regarding its configuration?  The defaults report a CPU temp of 52 C and the fan power is at 9%.

The start & end temps are 30 and 70 with a minimum power of 0.

Sorry I am so naive about this stuff!

**Reply by FlaMike** [reply]
Strange...  I switched to Fahrenheit & the fan power jumped to 28%.  Wow!

And the little bugger is speeding up & slowing down, possibly trying to keep the CPU around 120 F.  So awsome!  Thank you!!!

**Reply by rstrouse** [reply]
I can only refer to the npm audit developers as A-Holes.  For instance, that glob deprecation makes relative pathnames not work.

For the smart fan you can use different curves.  In the v6 version you can use a curve that increases power linearly starting with at one value and ending at another.  This means that for instance settings from 30-70 the fan will start slowly when the temp is 30c then ramp in a linear fashion to 100% at 70c.  I think though in your environment you should be at 100% below 65 so that it doesn't make it all the way to 70c.

You can also choose a logarithmic curve where it starts to ramp rapidly at the start temp then taper off when it reaches the top of the curve.   You can set the ramp rate to slow medium or fast.

The exponential ramp is the inverse of the logarithmic ramp where it starts slowly then ramps rapidly as heat builds up.

The minimum power is used to ensure the fan runs at a specified speed at a minumum.  So if the calculation only calculates to 5% and the minimum is to 10% it will run the fan at 10% power.

In the end I would try to keep the pi below 50 if possible but it will operate fine up to 85 where it will start throttling.  Honestly, these options are a bit of overkill but the mathnastics keep the brain fresh and provide toys for others.

**Reply by FlaMike** [reply]
Thank you for your detailed reply (not to mention your help earlier today and always!).

Once the updated REM software was  installed I was  able to switch to Fahrenheit so that's one fewer calculation needed on my part to keep things easy.  (We Americans are *so* out of  touch with most of the world.)  Anyway, at the moment I'm running the defaults--linear, start temp 86F, end temp 158F, min power 0.

When the little feller first came to life a couple of hours ago, the sun was beating down on the pad.  The reported Pi CPU temp with the above defaults running was about 121F.  Now the pad is in the shade and even though the ambient air temp is higher (about 97.7 if  the thermister is close), the Pi is about 116 and the fan is running at 22%.  Not too bad...

The case fan is not PWM.  It kicks on when the Pi is 90F, which means this time of year it's always on.   And unlike last week, there's now a Meanwell transformer in the box and I know it's generating some heat.  Overall, I'm OK with the current settings, but it's good to know there are plenty of toys in the sandbox!

Both fans are Noctuas, so I'm hoping they have a fairly long life.  I know the cheaper fans have less than stellar longevity & to me it's worth spending a couple of extra dollars for more reliable stuff that hopefully lasts longer.  That's especially true for equipment that's outside with plenty of heat & humidity.

Thanks again for all of your support throughout my adventure.  Have a super weekend & beyond!!

BTW, based on your explanation of the various fan curve options, I am thus far using the linear curve but set the end temp to 145F so the fan will be at 100% before the CPU reaches its throttling temperature--and do its best to keep the CPU below that  threshold.  I believe the Smart Fan, just like the case fan will probably be on most of the time.  That's fine with me & why they are part of the build.
